_posts/2017-08-25-introducing-sandboxfs.md

```diff
@@ -3,40 +3,46 @@ layout: posts
 words: we want actions to run within a sandbox.
 
-To better illustrate this point, let's consider an example: think of a C++
-`cc_library` target that specifies `foo.cc` as a source. This target contains a
+Let's consider an example. Think of a `cc_library` target that specifies
+`foo.cc` as a source. The `cc_library` rule contains a
 *compilation action* that turns the `foo.cc` source file into a `foo.o` object
-file. This action will run clang, will need to read `foo.cc`, will need to read
-the transitive closure of all headers required by `foo.cc`, will need to write
-`foo.o`, and may need to write some temporary files under `/tmp/`.
-
-Under these conditions, Bazel must ensure that the action only has read-only
-access to the input files we identified and write-only access to the output and
-temporary files we expect. The reason for this is because, unless these
-conditions are enforced, Bazel cannot know that the output file really came
-from those inputs: it may well be that the compiler went astray and read random
-files from the system, which means further builds would be inconsistent.
+file. This action will run clang, and it will need to:
+*  read `foo.cc`,
+*  read the transitive closure of all headers required by `foo.cc`
+*  write `foo.o`, and
+*  possibly write some temporary files under `/tmp/`.
+
+To be confident this build is correct, Bazel must ensure that the action has
+read-only access to the identified input files and write-only access
+to the output and temporary files we expect. Otherwise, Bazel cannot know if the
+compiler went astray and read random files from the system, making future builds
+inconsistent.
 
 # Current sandboxing techniques
 
 Today, Bazel uses different technologies to implement the sandboxing of
 actions. For example, on Linux, Bazel uses PID- and mount-namespaces, and on
-macOS uses symlinks and the `sandbox(7)` facility. (Note: sandboxing is
-disabled by default.)
+macOS Bazel uses symlinks and the `sandbox(7)` facility. Note that sandboxing
+is disabled by default.
 
-However, all these approaches have scalability and performance issues. The
-reason is that a typical build action requires hundreds, if not thousands, of
-files and directories to be mapped into the sandbox, and setting these up takes
-time. This is especially problematic on macOS: on this platform, Bazel must
+However, all these approaches have scalability and performance issues. A
+typical build action requires hundreds, if not thousands, of
+files and directories to be mapped into the sandbox. Setting these up takes
+time. On macOS this approach is especially problematic because Bazel must
 create thousands of symlinks every time it invokes an action, and this is slow.
 
-To make things worse, these approaches also have correctness issues: if
+To make things worse, these approaches also have correctness issues. If
 symlinks are used, some build tools will decide to extract the real path of
```

Comment:
"some build tools"  but Bazel is our only build tool!

Could sandboxfs be used by other build tools and so we want to be general, or here are we talking about the things in the toolchain that Bazel coordinates?

Reply:
"build tools" here refers to compilers, linkers, or whatever else actions invoke. Is there a better term we can use?

Reply:
Got it - things in the toolchain (binaries, tools, or libraries). So I believe "tools" is the term we want, but let's specify because it's such a generic word. I split the sentence because it was getting long:

"If symlinks are used, some tools (for example, some compilers or linkers) will decide to extract the real path of such symlinks and work off that path. These tools may end up "discovering" and consuming undeclared files that are siblings of the symlink's target."

Reply:
Done.